[WIP] pattern-matching: Cannot_flatten, maybe we can?

gasche opened the pull request:
This PR implements a new approach to "flattening" pattern-matching compilation (`do_for_multiple_match`) for source programs of the form `match foo, bar, baz with ...`. (The pattern-matching compiler is trying to avoid allocating a tuple if possible, that is if the patterns do not use the name of the tuple. The point of this PR is to change the details of how this is implemented.)

The goal of getting rid of the `Cannot_flatten` exception is to simplify reasoning on the codebase. It looks like we actually produce better code, but this was not the intent -- the examples where we see a different are relatively artificial and may not exist in real codebases.

This PR is only a Work-in-Progress, because the "simplify the reasoning" part is only half-achieved. We do not have `Cannot_flatten` anymore, so the control flow is simpler, but the correctness reasoning for the `flatten_*` functions has become more subtle: whereas we knew before that they would never encounter (and throw away) as-patterns, they can now reach as-patterns, but only in usage modes (default environments, provenance) where (I believe that) variables do not matter anyway, so ignoring them is fine.

(cc @maranget, @trefis, @Octachron)

gasche committed:
matching: document the behavior of precompile_or through an example

gasche committed:
Matching.explode_or_pat: handle the case where the argument has no id

gasche committed:
Match.do_for_multiple_match: we can always flatten

gasche committed:
matching: towards a correctness argument for flatten_* ignoring variables

Before we ignored as-patterns in the flatten_* functions because
as-patterns would either be half-simplified or raise Cannot_flatten
(in any case, never reach the flattening functions).

Now the reasoning is a bit more subtle: the only non-simple matrices
we flatten are used as "ghost" information (default environments,
provenance) where variables do not matter, only the shape of matched values.

gasche committed:
matching: more complete testsuite for do_for_multiple_match

Note: we now use -dlambda rather than -drawlambda, because otherwise
the output is much more verbose and difficult to read.

(-drawlambda is closed to the inner workings of the pattern-matching
compiler, but the simplification in -dlambda make the output much more
readable. They are also fairly predictable/non-surprising, so I think
that we can still easily understand what the compiler did from that
output.)

gasche commented:
We had a meeting today about this PR with @trefis and @maranget. @maranget is willing to get rid of exceptional control-flow, if the code really is simpler. @trefis or @Octachron proposed to review the PR (eventually) to get a third-party opinion (or two) on whether the change actually simplifies the code.

maranget commented on lambda/matching.ml:
```diff
@@ -567,9 +587,12 @@ let rec flatten_pat_line size p k =
+         where variables do not matter (default environments,
+         "provenance", etc.). *)
       flatten_pat_line size p k
+  | Tpat_var _ -> Patterns.omegas size :: k
```
Why not use a or-pattern add to the `Tpat_any` in place of this extra clause ?

gasche replied:
done

trefis commented on testsuite/tests/basic/patmatch_for_multiple.ml:
```diff
@@ -1,5 +1,5 @@
 (* TEST
-   flags = "-drawlambda"
+   flags = "-dlambda"
```
I'm not fond of this change. Granted the "raw lambda" output is more verbose, but we are sure about what we're looking at (the output of matching).
With `-dlambda`, `Simplif` will have run, and we're not just looking at the output of parmatch anymore. Admittedly this is also useful, but meh.

trefis commented on testsuite/tests/basic/patmatch_for_multiple.ml:
```diff
@@ -35,25 +25,179 @@ match (3, 2, 1) with
+      (exit 4 (makeblock 0 *match*/96 *match*/97 *match*/98)))
+   with (4 x/94) (seq (ignore x/94) 1)))
 - : bool = false
 |}];;
```
The comment above this test is now incorrect.

trefis commented on testsuite/tests/basic/patmatch_for_multiple.ml:
```diff
@@ -35,25 +25,179 @@ match (3, 2, 1) with
 |}];;
+
+(* Regression test for #3780 *)
+let f a b =
+  match a, b with
+  | ((true, _) as _g)
+  | ((false, _) as _g) -> ()
+[%%expect{|
+(let (f/106 = (function a/108 b/109 0))
+  (apply (field 1 (global Toploop!)) "f" f/106))
+val f : bool -> 'a -> unit = <fun>
+|}];;
```
There would be less noise in the expect-block if you removed the binding (i.e. if you had `(fun a b -> match a, b with ...)`).
That's also true for all the later tests in this file.

trefis reviewed (commented):
Some meta-comments about the tests.
I'll look more carefully at the code later.

gasche committed:
matching: avoid useless bindings in do_for_multiple_match

gasche commented:
I did the changes suggested by @trefis. To make the `-drawlambda` output less noisy I added some logic to avoid identifier-rebinding in `do_for_multiple_match`.

trefis commented on lambda/matching.ml:
```diff
@@ -1489,11 +1550,10 @@ and precompile_or ~arg_id (cls : Simple.clause list) ors args def k =
             let or_num = next_raise_count () in
             let new_patl = Patterns.omega_list patl in
             let mk_new_action ~vars =
-              Lstaticraise (or_num, List.map (fun v -> Lvar v) vars)
-            in
+              Lstaticraise (or_num, List.map (fun v -> Lvar v) vars) in
```
Please no.

trefis commented on lambda/matching.ml:
```diff
@@ -320,9 +308,41 @@ end = struct
+             compile a tuple pattern [match e1, .. en with ...]
+             without allocating the tuple [(e1, .., en)].
+          *)
+          let env = List.map (fun id -> id, Ident.rename id) vars in
```
Not all of the fresh new ids are being used, so you could delay the call to `Ident.rename` until you know that you're going to be using the id.
Also, I think the code would be easier to read without one more `env` variable with a different meaning from other `env`s in this file. (I know the name was there before your PR, but let's take the opportunity to change it!)

gasche replied:
I'm trying to apply your on-demand-rename suggestion:
- bad news: the resulting code is even more complex than the current one
- good news: this forced me to write a comment to explain what the function does in the first place, and I realized that we can probably simplify its interface a bit (it does not need to take the `patl` as part of the input as it ignores them, and in particular does not freshen any variables they would bind in the action, which is fine as the function is only called with omega-patterns there!).

gasche replied:
I pushed commits corresponding to those changes, let me know what you think.

trefis reviewed (commented):
Some further nitpicking, but otherwise the code looks correct to me.

trefis commented:
PS: What about the changes entry?

gasche committed:
add a comment to explain what Simple.explode_or_pat does

gasche committed:
Matching.Simple.explode_or_pat: on-demand generation of fresh identifiers

This change was suggested by Thomas Refis during code review.

gasche committed:
matching: simplify the interface of Simple.explode_or_pat

gasche commented:
@trefis I believe that I have addressed your comments (although for fresh-id generation you might think that the cure is worse than the disease, with the resulting code more complex; let me know and I can always revert), and now the CI is green. Hint hint.

gasche commented:
(recently I wondered about which of the pattern-matching PR are still in-flight; this one is)

trefis commented on lambda/matching.ml:
```diff
@@ -260,15 +263,11 @@ module Simple : sig
+    arg:lambda ->
+    Half_simple.pattern ->
     mk_action:(vars:Ident.t list -> lambda) ->
     vars:Ident.t list ->
```
Coming back to this code after some months, I couldn't remember straightaway what `vars` was, reading its implementation I could venture a guess, which I confirmed by looking at the (sole) call-site of this function.

At the said call-site, there's the following comment just above `vars` which I think could be moved up here:
> (* bound variables of the or-pattern and used in the orpm actions *)

gasche replied:
We might want to try to wait 3 months on every PR to see if the code still makes sense :-) Thanks for the excellent nitpick.

gasche replied:
Rebased. I renamed `vars` into `patbound_action_vars`, is that clearer? I did not end up moving the comment around due to decision fatigue (on the function signature or implementation? document just this parameter or also the rest?).

trefis reviewed (approved):
Sorry for the long delay, I'm satisfied with current state of things.

I did voice one nitpick, which hopefully you won't mind addressing since you need to rebase anyway to handle the conflict in Changes.

gasche committed:
matching: clarify the ~vars argument of Simple.explode_or_pat

(suggested by Thomas Refis' review)

trefis commented on Changes:
```diff
@@ -19,6 +19,9 @@ Working version
 ### Internal/compiler-libs changes:
 
+- #9650: keep refactoring the pattern-matching compiler
+  (Gabriel Scherer, review by Thomas Refis
```
You're missing a closing parenthesis.

Octachron commented on testsuite/tests/basic/patmatch_for_multiple.ml:
```diff
@@ -3,57 +3,197 @@
+[%%expect {|
+(function a/144 b/145
+  (if a/144
+    (let (x/146 =a a/144 _p/147 =a (makeblock 0 a/144 b/145))
```
Isn't the tuple being allocated here?

gasche replied:
It's subtle. The declaration shows up, but it is an unused alias, so it is removed by simplification: it is absent in the `-dlambda` output. I think that I wrote the original tests with `-dlambda`, and wrote the comments accordingly, then someone (my bet is on @trefis) asked for `-drawlambda` output, and I missed this small surprise. Thanks for noticing!

The testsuite is a bit hard to make sense of today, because it mostly shows things that are handled in the exact same way in all cases. Before the PR, it would show fairly different results; in particular, when an allocation happens due to Cannot_flatten, it is at the top of the matching (before any conditional test), and always used as the match scrutinee, so never erasable. Here we push the allocations within the branch, and in some cases they get erased by simplification (which is the point of trying harder to flatten).

The test that you are looking at is a test that was always flattenable. In particular, ocaml 4.10 and this PR produce the same `-dlambda` and the same `-drawlambda` output. The "outside, tuplist" test below is similar. But this is not the case for the "inside, tuplist" test below it, that now only allocates in one branch and not in the other (but the allocation shows in the `-drawlambda` output.)

It is not good that the test output is confusing in this way, but I am not sure how to fix it. We could:
- add a comment in the test
- revert to -dlambda reference output
- change the code generator to not generate a binding for a variable that is not used in the action anyway

I think I will start with just a comment, but I am not confident that this is the best approach. Any advice?

gasche replied:
Regarding the testsuite:  my point is that this test is not so useful to understand the behavior today, but rather to:
- show up as interesting diffs in the commit that implements the change of behavior in the PR
- test the space of various situations, and serve as regression tests if we evolve this part of the code in the future

It sounds believable to me that the same "differences between versions" that we had with Cannot_flatten could re-appear with another implementation; then we could decide whether we accept them or not, but at least we have the corner-case differences in the testsuite.

Octachron replied:
The tuplist examples avoid the allocation on the first branch, aren't they? 

I would be happy with comments that match the result of the tests. If I am not mistaken, with the rawlambda output, we are testing that the tuple allocation happen inside the branch, when needed, and independently of the presence of an exploded-or.

gasche replied:
Before this PR (for example in 4.10), the "outside" test would avoid the allocation in the first branch, but the "inside" test would raise Cannot_flatten and allocate in both cases. After the PR, both tests avoid the allocation in the first branch.

Thanks for the suggestion on how to write the comment.

gasche replied:
I rephrased the comment, and also I decided to inculde *both* `-drawlambda` and `-dlambda` output, to make the situation less perplexing in the first place.

Octachron replied:
This seems fine to me.

gasche committed:
testsuite: clarify the allocation behavior

(A nice catch of Florian Angeletti's review)

gasche committed:
testsuite: include both -dlambda and -drawlambda outputs
